docs: update tutorials (#565)
deepsense-ai/ragbits: docs/quickstart/llms.md

> # Quickstart: Working with Large Language Models

Reviewer: Also to be more diataxis oriented

```suggestion
# Tutorial: Question-Answering System with a Large Language Model
```

> Let's walk through a quick example of **basic question answering**. Specifically, let's build **a system for answering tech questions**, e.g. about Linux or iPhone apps.

Reviewer: According to https://diataxis.fr/tutorials/ It is recommended to start tutorials with 

*In this tutorial we will create and deploy a scalable web application. Along the way we will encounter containerisation tools and services.* 

```suggestion
In this tutorial, we will build **a system for answering tech questions**, such as those about Linux or iPhone apps. Along the way, we’ll explore key concepts like **Prompting**, **LiteLLM**, **observability**, and **evaluation**.
```

> Install the latest Ragbits via `pip install -U ragbits` and follow along.

Reviewer: So now this will become a tutorial, shpuldn't we assume that ragbits is already installed?

Author: i was following tutorials from dspy, they have this intro everywhere

## Configure environment

During development, we will use OpenAI's `gpt-4.1-nano` model. To authenticate, Ragbits will look into your `OPENAI_API_KEY`. You can easily swap this out for [other providers](../how-to/llms/use_llms.md) or [local models](../how-to/llms/use_local_llms.md).

> !!! tip "Recommended: Set up OpenTelemetry tracing to understand what's happening under the hood."

Reviewer: Again diataxis stuff 

![image](https://github.com/user-attachments/assets/18eb9133-8cb0-41ca-8f1a-c899725cb320)

So maybe we move this to the end of this tutorial then we will fulfill another checkpoint of diataxis which is **Encourage and permit repetition**:
1. User should run entire pipeline by itself
2. Then they set up OpenTelemetry and observe everything from a different viewpoint

    OpenTelemetry is an LLMOps tool that natively integrates with Ragbits and offer explainability and experiment tracking. In this tutorial, you can use OpenTelemetry to visualize prompts and optimization progress as traces to understand the Ragbits' behavior better. Check the full setup guide [here](../how-to/audit/use_tracing.md/#using-opentelemetry-tracer).

## Exploring Ragbits components

> You can always prompt the LLM directly via `llm.generate("prompt")` or `llm.generate([...])`. However, Ragbits gives you a better way to define your functions.

Reviewer: This breaks **Ignore options and alternatives** I assume this is for a beginner, so should we distract them in this way?

> The simplest way to define a prompt in Ragbits is to create a class that inherits from the [`Prompt`](../api_reference/core/prompt.md) class.

Reviewer: I'd use recommended word as `simplest` again implies that maybe `harder` is better etc.

```suggestion
The recommended way to define a prompt in Ragbits is to create a class that inherits from the [`Prompt`](../api_reference/core/prompt.md) class.
```

```python
from pydantic import BaseModel
from ragbits.core.prompt import Prompt

class QuestionAnswerPromptInput(BaseModel):
    question: str

class QuestionAnswerPromptOutput(BaseModel):
    answer: str

class QuestionAnswerPrompt(Prompt[QuestionAnswerPromptInput, QuestionAnswerPromptOutput]):
    system_prompt = """
    You are a question answering agent. Answer the question to the best of your ability.
    """
    user_prompt = """
    Question: {{ question }}
    """
```

In order to run this prompt, initilize LLM client and call `generate` method.

```python
from ragbits.core.llms import LiteLLM

async def main() -> None:
    llm = LiteLLM(model_name="gpt-4.1-nano", use_structured_output=True)
    prompt = QuestionAnswerPrompt(QuestionAnswerPromptInput(
        question="What are high memory and low memory on linux?",
    ))
    response = await llm.generate(prompt)
    print(response.answer)

if __name__ == "__main__":

>     asyncio.run(main())

Reviewer: Missing `import asyncio`

if __name__ == "__main__":
    asyncio.run(main())
```

```text
High memory on Linux refers to the part of RAM above the addressable limit for a 32-bit kernel, often above 1GB
or 4GB depending on architecture, which requires special handling to access. Low memory is the portion within
the addressable range for the kernel, generally below these limits, and is directly accessible by the system.
```

Interestingly, asking for reasoning can make the output answer shorter in this case. Is this a good thing or a bad thing? It depends on what you need: there's no free lunch, but Ragbits gives you the tools to experiment with different strategies extremely quickly.

## Evaluation

Before we get to evaluations, Ragbits provides evaluations only for 2-layer components, such as [`DocumentSearch`][ragbits.document_search.DocumentSearch] or [`Agent`][ragbits.agents.Agent]. To run the evaluations on LLM, you must use them through the [`Agent`][ragbits.agents.Agent] object.

```python
from ragbits.agents.types import QuestionAnswerAgent

llm = LiteLLM(model_name="gpt-4.1-nano", use_structured_output=True)
responder = QuestionAnswerAgent(llm=llm, prompt=CoTQuestionAnswerPrompt)
```

> To measure the quality of your Ragbits system, you need a bunch of input values, like questions for example, and a metric that can score the quality of an output from your system. Metrics vary widely. Some metrics need ground-truth labels of ideal outputs, e.g. for classification or question answering. Other metrics are self-supervised, e.g. checking faithfulness or lack of hallucination.

Reviewer: Again I'd keep it short and simple and straight to the point.

I'd remove explanation of metrics, this is not a point for the user to think what metrics exists here. Maybe just put a link that describes different metrics

```suggestion
To evaluate the quality of our QA system, we need two things:

1. A dataset of input questions (and ideally their correct answers).
2. A metric that can judge the quality of the system's responses.
```

> Let's load a dataset of questions and their ground truth answers. Since we started this tutorial with the goal of building a system for answering Tech questions, we obtained a bunch of questions and their correct answers from the [RAG-QA Arena](https://arxiv.org/abs/2407.13998) dataset.

Reviewer: again I'll keep it as short as possible

```suggestion
Since our goal is to build a system for answering technical questions, we’ll use examples from the [RAG-QA Arena](https://arxiv.org/abs/2407.13998) dataset.
```

```python
from ragbits.core.sources import WebSource
from ragbits.evaluate.dataloaders.question_answer import QuestionAnswerDataLoader

async def main() -> None:
    source = WebSource(url="https://huggingface.co/dspy/cache/resolve/main/ragqa_arena_tech_examples.jsonl")
    dataloader = QuestionAnswerDataLoader(
        source=source,

>         split="data[:500]",

Reviewer: Isn't 500 too much? 

* It is costly
* It runs for a long time

        question_key="question",
        answer_key="response",
    )
    dataset = await dataloader.load()
    print(dataset[0])

if __name__ == "__main__":
    asyncio.run(main())
```

```python
QuestionAnswerData(
    question="why igp is used in mpls?",
    reference_answer="An IGP exchanges routing prefixes between gateways/routers.  \nWithout a routing protocol, you'd have to configure each route on every router and you'd have no dynamic updates when routes change because of link failures. \nFuthermore, within an MPLS network, an IGP is vital for advertising the internal topology and ensuring connectivity for MP-BGP inside the network.",
    reference_context=None
)
```

> What kind of metric can suit our question-answering task? There are many choices, but since the answers are long, we may ask: How well does the system response cover all key facts in the gold response? And the other way around, how well is the system response not saying things that aren't in the gold response?

Reviewer: Plain text is little bit harder to read than clear bullets.

Again diataxis is all again no choice and no alternatives - `There are many choices`

```suggestion
So, what kind of metric is suitable for this QA task?

Since our answers are fairly long, we want a metric that evaluates two things:
- Does the system response cover the key facts from the reference answer?
- Does it avoid adding extra information that isn’t in the reference?
```

> That metric measures essentially an answer correctness, so let's load a `QuestionAnswerAnswerCorrectness` metric from Ragbits. This metric is actually implemented as a very simple Ragbits module using whatever LLM we're working with.

Reviewer: Would remove unnecessary details, If we have an API link that we can send users to would place it here

```suggestion
Ragbits includes a built-in metric for this called `QuestionAnswerAnswerCorrectness`.
```

```python

> from ragbits.agents.types import QuestionAnswerAgent

Reviewer: Duplicated import

from ragbits.evaluate.metrics.question_answer import QuestionAnswerAnswerCorrectness
from ragbits.evaluate.pipelines.question_answer import QuestionAnswerResult

async def main() -> None:
    llm = LiteLLM(model_name="gpt-4.1-nano", use_structured_output=True)
    responder = QuestionAnswerAgent(llm=llm, prompt=CoTQuestionAnswerPrompt)
    response = await responder.run(QuestionAnswerPromptInput(question=dataset[0].question))

    judge = LiteLLM(model_name="gpt-4.1")
    metric = QuestionAnswerAnswerCorrectness(judge)
    score = await metric.compute([
        QuestionAnswerResult(

>             question=dataset[0].question,

Reviewer: Here I need to know that I have to reuse dataset from previous iteration

            reference_answer=dataset[0].reference_answer,
            predicted_result=response,
        )
    ])
    print(score)

if __name__ == "__main__":
    asyncio.run(main())
```

```text
{'LLM_based_answer_correctness': 1.0}
```

> For evaluation, you could use the metric above in a simple loop and just average the score. But for nice parallelism and utilities, we can rely on `Evaluator`.

Reviewer: Alternatives are not recommended in tutorials. Also please provide a link to Evaluator API reference

```suggestion
To streamline the evaluation process with parallelism we can rely on `Evaluator`.
```

```python
from ragbits.agents.types import QuestionAnswerAgent
from ragbits.evaluate.evaluator import Evaluator
from ragbits.evaluate.metrics import MetricSet
from ragbits.evaluate.pipelines.question_answer import QuestionAnswerPipeline

evaluator = Evaluator()
results = await evaluator.compute(
    dataloader=dataloader,
    pipeline=QuestionAnswerPipeline(responder),
    metricset=MetricSet(metric),
)
print(results.metrics)
```

```text
{'LLM_based_answer_correctness': 0.68}
```

## Conclusions

> So far, we built a very simple LLM workflow using chain-of-thought for question answering and evaluated it on a small dataset.

Reviewer: ```suggestion
In this tutorial, we built a very simple LLM workflow using chain-of-thought for question answering and evaluated it on a small dataset.
```

> Can we do better? In the next guide, we will build a retrieval-augmented generation (RAG) pipeline in Ragbits for the same task. We'll see how this can boost the score substantially, then we'll use Ragbits optimizer to tune the retrieval parameters, raising our scores even more.

Reviewer: bullets points, also maybe API reference links?

```suggestion
Can we do better? In the next guide, we will:
- build a retrieval-augmented generation (RAG) pipeline in Ragbits for the same task and see how this can boost the score substantially
- Use Ragbits optimizer to tune the retrieval parameters, raising our scores even more.
```
